packages/vite/src/build.ts

```diff
@@ -0,0 +1,96 @@
+import type { Plugin } from 'vite';
+import { join, dirname } from 'path';
+import { createHash } from 'crypto';
+import { fork } from 'child_process';
+import type { AddonMeta } from '@embroider/core';
+import { ResolverLoader } from '@embroider/core';
+import { existsSync, readdirSync, readFileSync, rmSync, writeFileSync, rmdirSync, copyFileSync } from 'fs';
+import { mkdirpSync } from 'fs-extra';
+
+const cwd = process.cwd();
+const embroiderDir = join(cwd, 'node_modules', '.embroider');
+const cacheKeyPath = join(embroiderDir, 'cache-key.json');
+
+export const lockFiles = ['pnpm-lock.yaml', 'package-lock.json', 'yarn.lock', 'package.json'];
+
+function getCacheKey(file: string) {
+  if (existsSync(cacheKeyPath)) {
+    return JSON.parse(readFileSync(cacheKeyPath).toString())[file];
+  }
+  return null;
+}
+
+function updateCacheKey(file: string, key: string | null) {
+  let json: Record<string, string | null> = {};
+  if (existsSync(cacheKeyPath)) {
+    json = JSON.parse(readFileSync(cacheKeyPath).toString());
+  }
+  json[file] = key;
+  writeFileSync(cacheKeyPath, JSON.stringify(json));
+}
+
+function computeCacheKeyForFile(file: string) {
+  if (existsSync(file)) {
+    const fileBuffer = readFileSync(file);
+    const hashSum = createHash('sha256');
+    hashSum.update(fileBuffer);
+
+    return hashSum.digest('hex');
+  }
+  return null;
+}
+
+export function emberBuild(): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const child = fork('./node_modules/ember-cli/bin/ember', ['build']);
+    child.on('exit', code => (code === 0 ? resolve() : reject()));
+  });
+}
+
+export async function buildIfFileChanged(path: string | null | undefined): Promise<void> {
+  if (path && (lockFiles.includes(path) || path === 'app/index.html')) {
+    const key = computeCacheKeyForFile(path);
+    if (key !== getCacheKey(path)) {
+      console.log(path + ' change requires rebuild, rebuilding...');
+      await emberBuild();
+      updateCacheKey(path, key);
+    }
+  }
+}
+
+export function build(): Plugin {
+  let resolverLoader = new ResolverLoader(process.cwd());
+  const engine = resolverLoader.resolver.options.engines[0];
+  return {
+    name: 'embroider-builder',
+    enforce: 'pre',
+    writeBundle(options) {
+      engine.activeAddons.forEach(addon => {
+        const pkg = resolverLoader.resolver.packageCache.ownerOfFile(addon.root);
+        if (!pkg) return;
```

Comment:
If you do `!pkg?.isV2Addon()` you won't need to cast `pkg.meta` below.

Reply:
Will try it